### PR TITLE
fix(html-mismatch): Fixes warning in browser console

### LINKED ISSUE #1934: Expected server HTML to contain a matching <main> in <div>
The title (Expected server HTML to contain a matching ...) is the error message that shows after executing the very first command `yarn create redwood-app./appname`. Note that this message with the stacktrace is shown in the browser console:

image

---

**Copy of the text on the above image:**

```
index.js:1 Warning: Expected server HTML to contain a matching <main> in <div>.
    in main (created by SplashPage)
    in SplashPage (created by RouterImpl)
    in RouterImpl (created by LocationProvider)
    in LocationProvider (created by Context.Consumer)
    in Location (created by Router)
    in Router (created by Routes)
    in Routes (created by App)
    in FlashProvider (created by RedwoodApolloProvider)
    in GraphQLHooksProvider (created by RedwoodApolloProvider)
    in ApolloProvider (created by ApolloProviderWithFetchConfig)
    in ApolloProviderWithFetchConfig (created by RedwoodApolloProvider)
    in FetchConfigProvider (created by RedwoodApolloProvider)
    in RedwoodApolloProvider (created by App)
    in FatalErrorBoundary (created by App)
    in App
```

### PR DESCRIPTION
**Pull request by dac09**
Fixes #1934 #1887

Fix done in the way suggested by @peterp 

### What?
Changes the prerender and entry logic, so it looks for a specific `redwood-app-prerendered` element, instead of checking for `childNodes`. 

The issue with `childNodes` was that blank spaces, and comments are considered nodes, which is why we would see the watning the the browser console.

### CODE REVIEW

**Comment by thedavidprice** [comment]
@dac09 For projects with their own `index.js` (that overrides the built-in), would this be a "breaking" change we need to communicate for their specific case?

**Comment by dac09** [comment]
@thedavidprice - I don't believe this is a breaking change, but definitely worth investigating. Let me try it out and I'll report back.

I'm not so confident of this approach to be honest, my preference is definitely to stick to a single rootElement.

**Comment by dac09** [comment]
OK @thedavidprice, confirmed that the change is not breaking, but if you have a custom web index, then you continue receiving the warning in the console. 

Would love Peter P's go ahead on this one, but what I'm not a fan of:
a) Behaviour is inconsistent between dev and prod. In dev, any html entered in ```<div id="redwood-app">``` is replaced, where as in prod it isn't, but the content is simply appended (I believe this was the desired behaviour right @peterp ?)
b) Behaviour when prerendering is inconsitent to behaviour without prerender. i.e. if I add <h4>Hi, I'm loading</h4> in the HTML, it will be replaced if I don't prerender. If I do prerender, it will remain.

I'm going to take your direction here, but my preference would have been to only look for the root element, and replace

```
const rootElement = document.getElementById('redwood-app')

if (rootElement.children?.length > 0) {
  ReactDOM.hydrate(<App />, rootElement)
} else {
  ReactDOM.render(<App />,  rootElement)
}
```

If you want manual loaders before the redwood app loads, you can have it outside the "redwood-app" block and it will remain unaffected

**Comment by peterp** [comment]
@dac09 Those inconsistencies definitely do suck. Let's go with the way you suggested instead!

File: packages/prerender/src/runPrerender.tsx [hunk 1]
```diff
@@ -67,7 +67,7 @@ export const runPrerender = async ({
   // This is set by webpack by the html plugin
   const renderOutput = indexContent.replace(
     '<server-markup></server-markup>',
```
Comment (peterp) [inline packages/prerender/src/runPrerender.tsx:69]: I actually thought we would remove `server-markup` and put `<div id="redwood-app-prerendered" />` in `index.html` and then insert/ replace that with `<div id="redwood-app-prerendered">${componentAsHtml}</div>`

**Comment by dac09** [comment]
> @dac09 Those inconsistencies definitely do suck. Let's go with the way you suggested instead!

@peterp updated, please take a look when you get a chance, dankeschone!

**Comment by dac09** [comment]
@thedavidprice verified backward compatibility, non-breaking, but still warns if you have a custom-web-index you generated earlier. Fix is as follows:

In **web/src/index.js**
```diff
-if (rootElement.hasChildNodes()) {
+if (rootElement.children?.length > 0) {
  ReactDOM.hydrate(<App />, rootElement)
} else {
  ReactDOM.render(<App />, rootElement)
}
```

**Review by peterp — APPROVED** [review]
🍰

**Pull request merged**